src/test/java/ArrayIndexedBagTest.java

```diff
@@ -0,0 +1,449 @@
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        bag.remove(1);
+        assertEquals(2, bag.size());
+    }
+
+    @Test
+    void removeByIndexSoleElementForIndexedBagIsEmpty() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.remove(0);
+        assertTrue(bag.isEmpty());
+    }
+
+    @Test
+    void removeByIndexNotInRangeOfIndexedBagThrowsException() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(101);
+        bag.add(103);
+        assertThrows(IndexOutOfBoundsException.class, () -> bag.remove(22));
+    }
+
+    @Test
+    void removeByNegativeIndexOfIndexedBagPropagatesException() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        assertThrows(IndexOutOfBoundsException.class, () -> bag.remove(-1));
+    }
+
+    @Test
+    void iteratorReturnsIterator() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        Iterator<Integer> it = bag.iterator();
+        assertNotNull(it);
+        assertTrue(it instanceof Iterator);
+    }
+
+    @Test
+    void iteratorHasNextOnNonEmptyIndexedBagReturnsTrue() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        Iterator<Integer> it = bag.iterator();
+        assertTrue(it.hasNext());
+        assertTrue(it.hasNext());
+    }
+
+    @Test
+    void iteratorHasNextOnEmptyIndexedBagReturnsFalse() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        Iterator<Integer> it = bag.iterator();
+        assertFalse(it.hasNext());
+    }
+
+    @Test
+    void iteratorHasNextAfterIteratingOverAllElementsOnIndexedBagReturnsFalse() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(99);
+        bag.add(99);
+        Iterator<Integer> it = bag.iterator();
+        it.next();
+        it.next();
+        it.next();
+        assertFalse(it.hasNext());
+    }
+
+    @Test
+    void iteratorHasNextAfterRemovingAllElementsFromIndexedBagReturnsFalse() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        bag.add(99);
+        bag.add(99);
+        bag.remove(0);
+        bag.remove(0);
+        bag.remove(0);
+        Iterator<Integer> it = bag.iterator();
+        assertFalse(it.hasNext());
+    }
+
+    @Test
+    void iteratorNextOnNonEmptyIndexedBagReturnsElement() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        bag.add(99);
+        Iterator<Integer> it = bag.iterator();
+        assertEquals(99, it.next());
+    }
+
+    @Test
+    void iteratorNextOnNonEmptyIndexedBagReturnsElementsInCorrectOrder() {
+        IndexedBag<Integer> bag = new ArrayIndexedBag<>();
+        for (int i = 0; i < 10; ++i) {
+            bag.add(i);
+        }
+        Iterator<Integer> it = bag.iterator();
+        for (int i = 0; i < 10; ++i) {
+            assertEquals(i, it.next());
+        }
+    }
```

Comment:
Two considerations I had:
1. Worth adding to arbitrary indices? Orderness was already tested above
2. Worth using `while(it.hasNext())` for 2nd for loop?